infrastructure/src/main/java/org/corfudb/infrastructure/remotecorfutable/DatabaseHandler.java

```diff
@@ -0,0 +1,472 @@
+    }
+
+    public void addTable(ByteString streamID) throws RocksDBException {
+        if (columnFamilies.containsKey(streamID)) {
+            return;
+        }
+        ColumnFamilyOptions tableOptions = new ColumnFamilyOptions();
+        tableOptions.optimizeUniversalStyleCompaction();
+        tableOptions.setComparator(BuiltinComparator.REVERSE_BYTEWISE_COMPARATOR);
+        ColumnFamilyDescriptor tableDescriptor = new ColumnFamilyDescriptor(streamID.toByteArray(), tableOptions);
+        ColumnFamilyHandle tableHandle;
+        try {
+            tableHandle = database.createColumnFamily(tableDescriptor);
+        } catch (RocksDBException e) {
+            log.error("Error in creating column family for table {}.", streamID.toString(DATABASE_CHARSET));
+            log.error("Cause of error: ", e);
+            tableOptions.close();
+            throw e;
+        }
+
+        ColumnFamilyOptions metadataOptions = new ColumnFamilyOptions();
+        metadataOptions.optimizeForPointLookup(METADATA_COLUMN_CACHE_SIZE);
+        ColumnFamilyDescriptor metadataDescriptor = new ColumnFamilyDescriptor(
+                getMetadataColumnName(streamID).toByteArray(), metadataOptions);
+
+        ColumnFamilyHandle metadataHandle;
+        try {
+             metadataHandle = database.createColumnFamily(metadataDescriptor);
+        } catch (RocksDBException e) {
+            log.error("Error in creating metadata column family for table {}.", streamID.toString(DATABASE_CHARSET));
+            log.error("Cause of error: ", e);
+            tableHandle.close();
+            try {
+                database.dropColumnFamily(tableHandle);
+            } catch (RocksDBException r) {
+                log.error("Error in dropping column family for table {}.", streamID.toString(DATABASE_CHARSET));
+                log.error("Cause of error: ", r);
+                throw e;
+            } finally {
+                metadataOptions.close();
+            }
+            throw e;
+        }
+
+        synchronized (columnFamilies) {
+         columnFamilies.put(streamID, tableHandle);
+         columnFamilies.put(getMetadataColumnName(streamID), metadataHandle);
+        }
+    }
+    //TODO: add aysnc API
+
+    /**
+     * This function provides an interface to read a key from the database.
+     * @param encodedKey The key to read.
+     * @param streamID The stream backing the database to read from.
+     * @return A byte[] containing the value mapped to the key, or null if the key is associated to a null value.
+     * @throws RocksDBException An error raised in scan.
+     */
+    public byte[] get(byte[] encodedKey, ByteString streamID) throws RocksDBException, DatabaseOperationException {
+        final byte[] returnVal;
+        if (!columnFamilies.containsKey(streamID)) {
+            throw new DatabaseOperationException("GET", "Invalid stream ID");
+        }
+        RocksIterator iter = database.newIterator(columnFamilies.get(streamID));
```

Comment:
you will have resource leaks here, please take a look at how this logic is implemented in PersistedStreamingMap without leaks

Reply:
I'm not sure where the resource leak is coming from. The only resources allocated are in the iterator, and the iterator is closed on error and termination.

Reply:
You had your chance to figure out that by yourself :) so let me enjoy dig into it :)
Probably the leak is not there, but it's hard to prove it by looking at the code.

- offtop: `returnVal = null;` not a great idea IMO. Your code would fail with NPE at any time in any other method which would try to get access into the array.
- offtop: `columnFamilies` - race condition could be possible if put and get happens at the same time in different threads then your method throws an exception
- I don't like the idea of throwing `RocksDBException`, and also if you re throw an exception then never write it to the log, otherwise you'd have multiple exceptions logged in the file on different levels which is very confusing and will consume a lot of resources.
-  `iter.status();` is used whether to throw an exception or return null, which so vague logic, and hard to develop any intuition and motivation why it's done that way. I would get rid of it completely. So that `if` condition needs to be refactored completely
- and finally `iter`, I hope this version of code looks better for you:
```
try(RocksIterator iter = database.newIterator(columnFamilies.get(streamID))) {
            iter.seek(encodedKey);

            if (!iter.isValid()) {
                try {
                    iter.status();
                } catch (RocksDBException e) {
                    //We've hit some error in iteration
                    log.error("Error in RocksDB get operation:", e);
                    throw e;
                }
                returnVal = null;
            } else {
                if (Arrays.equals(KeyEncodingUtil.extractEncodedKey(iter.key()),
                        KeyEncodingUtil.extractEncodedKey(encodedKey))) {
                    //This works due to latest version first comparator
                    returnVal = iter.value();
                } else {
                    //TODO: add read from log stream check
                    returnVal = null;
                }
            }
        }
```